crates/matrix-sdk-base/src/client.rs

```diff
@@ -643,22 +643,55 @@ impl BaseClient {
             };
 
             if let AnyGlobalAccountDataEvent::Direct(e) = &event {
+                let mut new_dms = HashMap::<&RoomId, HashSet<OwnedUserId>>::new();
                 for (user_id, rooms) in e.content.iter() {
                     for room_id in rooms {
-                        trace!(
-                            ?room_id, target = ?user_id,
-                            "Marking room as direct room"
-                        );
+                        new_dms.entry(room_id).or_default().insert(user_id.clone());
+                    }
+                }
 
-                        if let Some(room) = changes.room_infos.get_mut(room_id) {
-                            room.base_info.dm_targets.insert(user_id.clone());
-                        } else if let Some(room) = self.store.room(room_id) {
-                            let mut info = room.clone_info();
-                            if info.base_info.dm_targets.insert(user_id.clone()) {
-                                changes.add_room(info);
-                            }
+                let rooms = self.store.rooms();
+                let mut old_dms = rooms
+                    .iter()
```

Comment:
I feel like for such iterations over rooms, we need a `Store::rooms_iter()` API. Otherwise it's clearly a waste of resources. I am doing it now.

Reply:
Hmm, it's harder than expected to do that wit rustc stable. Not a blocker here.